### ISSUE TITLE
Fix OpenOCD integration with PlatformIO

### ISSUE THREAD

**Issue by jmadden173** [jlab-sensing/soil-power-sensor-firmware#1]
Currently the upload command gives the following error

```
(.venv) jtmadden@spruce:~/repos/jlab/soil-power-sensor-firmware/stm32 (HelloWorld) > pio run -t upload                                                                             [130]
Processing wioe5 (platform: https://github.com/jlab-sensing/platform-ststm32; board: WioE5; framework: stm32cube)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/WioE5.html
PLATFORM: ST STM32 (17.0.0+sha.1e8839c) > WioE5
HARDWARE: STM32WL55JC 48MHz, 64KB RAM, 256KB Flash
DEBUG: Current (stlink) On-board (stlink) External (cmsis-dap, jlink)
PACKAGES: 
 - framework-stm32cubewl @ 1.26.2+sha.77fa61b 
 - tool-dfuutil @ 1.11.0 
 - tool-dfuutil-arduino @ 1.11.0 
 - tool-ldscripts-ststm32 @ 0.2.0 
 - tool-openocd @ 3.1200.0 (12.0) 
 - tool-stm32duino @ 1.0.1 
 - toolchain-gccarmnoneeabi @ 1.70201.0 (7.2.1)
Warning! Cannot find the default startup file for `stm32wl55jc`. Ignore this warning if the startup code is part of your project.
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 12 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Checking size .pio/build/wioe5/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [          ]   0.1% (used 44 bytes from 65536 bytes)
Flash: [          ]   1.5% (used 3872 bytes from 262144 bytes)
Configuring upload protocol...
AVAILABLE: cmsis-dap, jlink, mbed, stlink
CURRENT: upload_protocol = stlink
Uploading .pio/build/wioe5/firmware.elf
xPack Open On-Chip Debugger 0.12.0-01004-g9ea7f3d64-dirty (2023-01-30-15:03)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
debug_level: 1

hla_swd
none separate

Error: read version failed
in procedure 'program'
** OpenOCD init failed **
shutdown command invoked

*** [upload] Error 1
============================================================================== [FAILED] Took 2.30 seconds ==============================================================================
```

Running the openocd command with the following reset scripts yields the following error message.

`reset.cfg`
```
transport select hla_swd

reset_config srst_only
```

**Error**
```
(.venv) jtmadden@spruce:~/.platformio/packages/tool-openocd/openocd/scripts > sudo ~/.platformio/packages/tool-openocd/bin/openocd -d3 -s /home/jtmadden/.platformio/packages/tool-openocd/openocd/scripts -f interface/stlink.cfg -c "transport select hla_swd" -f /home/jtmadden/repos/jlab/soil-power-sensor-firmware/stm32/reset.cfg -c "program {.pio/build/wioe5/firmware.elf}  verify reset; shutdown;"
xPack Open On-Chip Debugger 0.12.0-01004-g9ea7f3d64-dirty (2023-01-30-15:03)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
User : 3 2 options.c:52 configuration_output_handler(): debug_level: 3
User : 4 2 options.c:52 configuration_output_handler(): 
Debug: 5 2 configuration.c:33 add_script_search_dir(): adding /home/jtmadden/.platformio/packages/tool-openocd/openocd/scripts
Debug: 6 2 options.c:233 add_default_dirs(): bindir=/__w/openocd-xpack/openocd-xpack/build/linux-x64/application/bin
Debug: 7 2 options.c:234 add_default_dirs(): pkgdatadir=/__w/openocd-xpack/openocd-xpack/build/linux-x64/application/openocd
Debug: 8 2 options.c:235 add_default_dirs(): exepath=/home/jtmadden/.platformio/packages/tool-openocd/bin
Debug: 9 2 options.c:236 add_default_dirs(): bin2data=../openocd
Debug: 10 2 configuration.c:33 add_script_search_dir(): adding /root/.config/openocd
Debug: 11 2 configuration.c:33 add_script_search_dir(): adding /root/.openocd
Debug: 12 2 configuration.c:33 add_script_search_dir(): adding /home/jtmadden/.platformio/packages/tool-openocd/bin/../openocd/site
Debug: 13 2 configuration.c:33 add_script_search_dir(): adding /home/jtmadden/.platformio/packages/tool-openocd/bin/../openocd/scripts
Debug: 14 2 command.c:155 script_debug(): command - ocd_find interface/stlink.cfg
Debug: 15 2 configuration.c:88 find_file(): found interface/stlink.cfg
Debug: 16 2 command.c:155 script_debug(): command - adapter driver hla
Debug: 17 2 command.c:155 script_debug(): command - hla_layout stlink
Debug: 18 2 hla_interface.c:230 hl_interface_handle_layout_command(): hl_interface_handle_layout_command
Debug: 19 2 command.c:155 script_debug(): command - hla_device_desc ST-LINK
Debug: 20 2 hla_interface.c:217 hl_interface_handle_device_desc_command(): hl_interface_handle_device_desc_command
Debug: 21 2 command.c:155 script_debug(): command - hla_vid_pid 0x0483 0x3744 0x0483 0x3748 0x0483 0x374b 0x0483 0x374d 0x0483 0x374e 0x0483 0x374f 0x0483 0x3752 0x0483 0x3753 0x0483 0x3754 0x0483 0x3757
Debug: 22 2 command.c:155 script_debug(): command - transport select hla_swd
Debug: 23 2 hla_transport.c:191 hl_swd_transport_select(): hl_swd_transport_select
User : 24 3 options.c:52 configuration_output_handler(): hla_swdUser : 25 3 options.c:52 configuration_output_handler(): 
Debug: 26 3 command.c:155 script_debug(): command - ocd_find /home/jtmadden/repos/jlab/soil-power-sensor-firmware/stm32/reset.cfg
Debug: 27 3 configuration.c:88 find_file(): found /home/jtmadden/repos/jlab/soil-power-sensor-firmware/stm32/reset.cfg
Debug: 28 3 command.c:155 script_debug(): command - transport select hla_swd
Warn : 29 3 transport.c:276 jim_transport_select(): Transport "hla_swd" was already selected
Debug: 30 3 command.c:155 script_debug(): command - reset_config srst_only
User : 31 3 options.c:52 configuration_output_handler(): srst_only separate srst_gates_jtag srst_open_drain connect_deassert_srst
User : 32 3 options.c:52 configuration_output_handler(): 
Debug: 33 3 command.c:155 script_debug(): command - init
Debug: 34 3 command.c:155 script_debug(): command - target init
Debug: 35 3 command.c:155 script_debug(): command - target names
Debug: 36 3 target.c:1657 handle_target_init_command(): Initializing targets...
Warn : 37 3 adapter.c:140 adapter_init(): An adapter speed is not selected in the init scripts. OpenOCD will try to run the adapter at the low speed (100 kHz)
Warn : 38 3 adapter.c:143 adapter_init(): To remove this warnings and achieve reasonable communication speed with the target, set "adapter speed" or "jtag_rclk" in the init scripts.
Debug: 39 3 adapter.c:251 adapter_config_khz(): handle adapter khz
Debug: 40 3 adapter.c:215 adapter_khz_to_speed(): convert khz to adapter specific speed value
Debug: 41 3 hla_interface.c:111 hl_interface_init(): hl_interface_init
Debug: 42 3 hla_layout.c:84 hl_layout_init(): hl_layout_init
Debug: 43 3 adapter.c:215 adapter_khz_to_speed(): convert khz to adapter specific speed value
Debug: 44 3 adapter.c:219 adapter_khz_to_speed(): have adapter set up
Debug: 45 3 adapter.c:215 adapter_khz_to_speed(): convert khz to adapter specific speed value
Debug: 46 3 adapter.c:219 adapter_khz_to_speed(): have adapter set up
Info : 47 3 adapter.c:179 adapter_init(): clock speed 100 kHz
Debug: 48 3 openocd.c:134 handle_init_command(): Debug Adapter init complete
Debug: 49 3 command.c:155 script_debug(): command - transport init
Debug: 50 3 transport.c:219 handle_transport_init(): handle_transport_init
Debug: 51 3 hla_transport.c:143 hl_transport_init(): hl_transport_init
Error: 52 3 target.c:541 get_current_target(): BUG: current_target out of bounds
```

**NOTE:** This was after having to add *VID* and *PID* for `ST-Link V3PWR` to `interface/stlink.cfg`. Updated version looks like
```
# SPDX-License-Identifier: GPL-2.0-or-later

#
# STMicroelectronics ST-LINK/V1, ST-LINK/V2, ST-LINK/V2-1, STLINK-V3 in-circuit
# debugger/programmer
#

adapter driver hla
hla_layout stlink
hla_device_desc "ST-LINK"
hla_vid_pid 0x0483 0x3744 0x0483 0x3748 0x0483 0x374b 0x0483 0x374d 0x0483 0x374e 0x0483 0x374f 0x0483 0x3752 0x0483 0x3753 0x0483 0x3754 0x0483 0x3757

# Optionally specify the serial number of ST-LINK/V2 usb device.  ST-LINK/V2
# devices seem to have serial numbers with unreadable characters.  ST-LINK/V2
# firmware version >= V2.J21.S4 recommended to avoid issues with adapter serial
# number reset issues.
# eg.
# adapter serial "\xaa\xbc\x6e\x06\x50\x75\xff\x55\x17\x42\x19\x3f"
```

*Edit:* My best guess is that the xpack `openocd` does not work when flashing with SWD using ST-Link. The STM build of `openocd` works on flashing.

**Comment by jmadden173** [comment]
Uploading through *STM32CubeProgrammer* is still functional and will be used for the time being until a fix is found

**Comment by jmadden173** [comment]
The following resources may be helpful in resolving the issue:

[OpenOCD: enabling SWD on STLINK-V3](https://electronics.stackexchange.com/questions/506468/openocd-enabling-swd-on-stlink-v3)
[OpenOCD](https://github.com/STMicroelectronics/OpenOCD)

We may have to set our board to use the STM version of OpenOCD

**Comment by jmadden173** [comment]
With `STMicroelectronics/OpenOCD` was able to successfully flash the firmware. Now I need to figure out how to create the platformio tool library so that it can be integrated with the toolchain.

**Comment by jmadden173** [comment]
Reopening issue until debugging is implemented with OpenOCD.

The current status is flashing has been fixed by #20 

The debugging is close. I believe that it is an issue with the device not being held in reset by OpenOCD which is what the `gdb` debugger expects. I think the default `openocd` command expects JTAG is used with the hardware reset pin. Something need to be modified in this command needs to 

```
openocd -d3 -s /home/jtmadden/.platformio/packages/tool-openocd/openocd/scripts -f interface/stlink.cfg -c "transport select hla_swd" -f /home/jtmadden/repos/jlab/soil-power-sensor-firmware/stm32/reset.cfg -c "program {.pio/build/wioe5/firmware.elf}  verify reset; shutdown;"
```

Specifically
```
"program {.pio/build/wioe5/firmware.elf}  verify reset; shutdown;"
```

**NOTE:** This was just a example command that I copied from above. The actual command called may differ.

I'm also not sure exactly where `openocd` is called. Most likely will be in `platform-ststm32` repo.
